src/conn/conn_layered.c

```diff
@@ -1276,3 +1276,172 @@ __wt_disagg_advance_checkpoint(WT_SESSION_IMPL *session, bool ckpt_success)
+        WT_ERR_NOTFOUND_OK(version_cursor->next(version_cursor), true);
+        if (ret == WT_NOTFOUND) {
+            if (key->size > 0 && upds != NULL)
+                WT_ERR(__layered_move_updates(cbt, key, upds));
+            else
+                ret = 0;
+            break;
+        }
+
+        WT_ERR(version_cursor->get_key(version_cursor, tmp_key));
+        WT_ERR(__wt_compare(session, CUR2BT(cbt)->collator, key, tmp_key, &cmp));
+        if (cmp != 0) {
+            WT_ASSERT(session, cmp <= 0);
+
+            if (upds != NULL)
+                WT_ERR(__layered_move_updates(cbt, key, upds));
+
+            upds = NULL;
+            prev_upd = NULL;
+            WT_ERR(__wt_buf_set(session, key, tmp_key->data, tmp_key->size));
+        }
+
+        WT_ERR(version_cursor->get_value(version_cursor, &tw.start_txn, &tw.start_ts,
+          &tw.durable_start_ts, &tw.stop_txn, &tw.stop_ts, &tw.durable_stop_ts, &type, &prepare,
+          &flags, &location, value));
+        /* We shouldn't seen any prepared updates. */
+        WT_ASSERT(session, prepare == 0);
+
+        /*
+         * We assume the updates returned will be in timestamp order. TODO: ensure the version
+         * cursor don't return duplicate values to honor the timestamp ordering.
+         */
+        if (prev_upd != NULL) {
+            WT_ASSERT(session,
+              tw.stop_txn <= prev_upd->txnid && tw.stop_ts <= prev_upd->start_ts &&
+                tw.durable_stop_ts <= prev_upd->durable_ts);
+            WT_ASSERT(session,
+              tw.start_txn <= prev_upd->txnid && tw.start_ts <= prev_upd->start_ts &&
+                tw.durable_start_ts <= prev_upd->durable_ts);
+            if (tw.stop_txn != prev_upd->txnid || tw.stop_txn != prev_upd->start_ts ||
+              tw.durable_stop_ts != prev_upd->durable_ts)
+                WT_ERR(__wt_upd_alloc_tombstone(session, &tombstone, NULL));
+        } else if (WT_TIME_WINDOW_HAS_STOP(&tw))
+            WT_ERR(__wt_upd_alloc_tombstone(session, &tombstone, NULL));
+
+        WT_ERR(__wt_upd_alloc(session, value, WT_UPDATE_STANDARD, &upd, NULL));
+        upd->txnid = tw.start_txn;
+        upd->start_ts = tw.start_ts;
+        upd->durable_ts = tw.durable_start_ts;
+
+        if (tombstone != NULL) {
+            tombstone->txnid = tw.stop_txn;
+            tombstone->start_ts = tw.start_ts;
+            tombstone->durable_ts = tw.durable_start_ts;
+            tombstone->next = upd;
+
+            if (prev_upd != NULL)
+                prev_upd->next = tombstone;
+            else {
+                prev_upd = upd;
+                upds = tombstone;
+            }
+            tombstone = NULL;
+            upd = NULL;
+        } else {
+            if (prev_upd != NULL)
+                prev_upd->next = upd;
+            else {
+                prev_upd = upd;
+                upds = upd;
+            }
+            upd = NULL;
+        }
+    }
+
+err:
+    if (tombstone != NULL)
```

Comment:
Close cursors?

Reply:
That's only the stable cursor though, the version cursor is something we open.

Reply:
oh, good catch.